src/ethereum-json-rpc-client/src/state_machine_tests_client.rs

```diff
@@ -0,0 +1,102 @@
+use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
+
+use anyhow::Context;
+use candid::{CandidType, Deserialize};
+use ic_canister_client::StateMachineCanisterClient;
+use jsonrpc_core::{Request, Response};
+use serde::Serialize;
+use serde_bytes::ByteBuf;
+
+use crate::Client;
+
+impl Client for StateMachineCanisterClient {
+    fn send_rpc_query_request(
+        &self,
+        request: Request,
+    ) -> Pin<Box<dyn Future<Output = anyhow::Result<Response>> + Send + Sync>> {
+        let client = self.clone();
+
+        Box::pin(async move {
+            log::trace!("CanisterClient - sending 'http_request'. request: {request:?}");
+
+            let args = HttpRequest::new(&request)?;
+
+            let http_response: HttpResponse = client
+                .query("http_request", (args,))
+                .await
+                .context("failed to send RPC request")?;
+
+            let response = serde_json::from_slice(&http_response.body)
+                .context("failed to deserialize RPC request")?;
+
+            log::trace!("response: {:?}", response);
+
+            Ok(response)
+        })
+    }
+
+    fn send_rpc_update_request(
+        &self,
+        request: Request,
+    ) -> Pin<Box<dyn Future<Output = anyhow::Result<Response>> + Send + Sync>> {
+        let client = self.clone();
+
+        Box::pin(async move {
+            log::trace!("CanisterClient - sending 'http_request_update'. request: {request:?}");
+
+            let args = HttpRequest::new(&request)?;
+
+            let http_response: HttpResponse = client
+                .update("http_request_update", (args,))
+                .await
+                .context("failed to send RPC request")?;
+
+            let response = serde_json::from_slice(&http_response.body)
+                .context("failed to deserialize RPC request")?;
+
+            log::trace!("response: {:?}", response);
+
+            Ok(response)
+        })
+    }
+}
+
+/// The important components of an HTTP request.
+#[derive(Clone, Debug, CandidType)]
+struct HttpRequest {
```

Comment:
This structures are slightly different than the ones in the evm-canister so they cannot be shared